### DISCUSSION TITLE
Is it possible to enforce `public` functions to return only immutable data structures?

### DISCUSSION THREAD

**Discussion by dumanqm** [LemonAppDev/konsist#1345] [Ask a Question] [answered]
### Wanted end result.

Hi team 👋

Since discovering the Konsist library, I’ve been using it across our repositories, and it's been a great tool!

I have a question and would appreciate your help in figuring out how to use the library further. I'd like to create a new Konsist function to check return types for public functions, ensuring that their return types are immutable data structures. The goal is to enforce that mutability is handled only within the class itself, and no mutable objects are exposed to outside classes.

Since we’re using `StateFlow`, `Compose`, and other immutability-dependent frameworks, it’s important for us to guarantee that public/internal functions and properties only provide immutable data structures.

### What you tried so far.

I’ve tried the following test case, but halfway through, I’m not sure what should be checked to ensure immutability.

```
    @Test
    fun `Test GIVEN all public functions WHEN checking the return type THEN ensure it is immutable data structure`() {
        rpaCoreScope
            .functions()
            // only focus on public and internal functions
            .withoutPrivateModifier()
            // Filter only functions which has return types
            .withReturnValue()
            // Skip the function which has Unit, Int, String, etc. return types
            .withoutReturnType { it.isKotlinBasicType }
            .forEach { function ->
                // What should be checked here?
            }
    }
```

**Comment by igorwojda** [comment, marked as answer]
Thx for the good feedback and info about usage :-)

All of the Kotlin Mutable data structures have `"Mutable"` in the type name e.g.

```
MutableCollection
MutableList
MutableMap
MutableSet
```

This seems to be industry accepted pattern that is also followed by coroutines...

```
MutableStateFlow
```

...and Android.
```
MutableState
MutableLiveData
```

If these are the only types you want to avoid then you can check if type name contains a `"Mutable"` string:

```koltin
rpaCoreScope
    .functions()
    // only focus on public and internal functions
    .filter { it.hasPublicOrDefaultModifier || it.hasInternalModifier }
    // No "Mutable" string in return type name
    .assertFalse {
        it.returnType?.text?.contains("Mutable")
    }
```

Notice that this example also improves filtering of public and internal functions.

If you need to exclude additional types you can update `assertFalse` block code to verify more types

```kotlin
val forbiddenTypes = listOf("MyType1", "MyType2")
...
.assertFalse { function ->
    val returnTypeText = function.returnType?.text.orEmpty()
    returnTypeText.contains("Mutable") || forbiddenTypes.any { it == returnTypeText } 
}
```

BTW it looks like we can improve Konsist API to make above tests easier to write:
- Add `representsTypeOf` method to return type to improve API https://lemonappdev.atlassian.net/browse/KON-635
- Add `isMutableType` property https://lemonappdev.atlassian.net/browse/KON-636

**Reply by dumanqm** [reply]
**@igorwojda, thanks for your quick response! 💯**

Two small notes:

1. I believe the `withoutPrivateModifier()` already works similarly to `.filter { it.hasPublicOrDefaultModifier || it.hasInternalModifier }`. I’m not sure which one is better performance-wise, but functionality-wise, they give the same results, right?
   
2. Is it possible to have the same test case for all `public` and `internal` properties?

To further clarify, we have backing fields in our classes as follows:

```kotlin
private val _uiState: MutableStateFlow = MutableStateFlow(HomeUiState())
val uiState: StateFlow = _uiState.asStateFlow()
```

However, sometimes people forget to specify the explicit type (`StateFlow`) or use the `asStateFlow()` extension function, as shown below:

```kotlin
val uiState = _uiState
```

In such cases, it would be great to ensure that the public `uiState` value is immutable (`StateFlow`) instead of the mutable (`MutableStateFlow`) version.

I tried the following example to check for this, but could you please confirm if the logic inside `assertFalse {}` is correct or if it could be further improved? 🙏

```kotlin
rpaCoreScope
    .properties()
    .withoutPrivateModifier()
    .assertFalse {
        it.type?.text?.contains("Mutable")
    }
```

**Reply by igorwojda** [reply]
1. Performance wise this makes no difference, IMO however it's better to express what you want to filter (positive statement) rather than what you want to exclude (negative stateent) - this makes tests are more readable. Also notice that this can case bugs like  in `2.` example - your test would ignore properties with `internal` visibility modifier which is undesired.

I would do this
2.
```kotlin
rpaCoreScope
    .properties()
    .withoutPublicOrDefaultModifier()
    .assertFalse {
        it.text?.contains("asStateFlow")
    }
```

or perhaps this to cover both cases:

```kotlin
...
.assertTrue {
        it.type?.name?.contains("Mutable") || it.text?.contains("asStateFlow")
    }
```

BTW it's best to use evaluate expression to see what data is available in `assert` block https://docs.konsist.lemonappdev.com/features/debug-konsist-test#evaluate-expression-debugger-window
